### PR TITLE
bp-messages: don't overflow when counting rewarded messages

### LINKED ISSUE #12907: [RWF] bp-messages: full-width nonce ranges overflow relayer reward calculation
## Summary

The `calc_relayers_rewards` function in bridges/primitives/messages/src/lib.rs (line 521) computes the number of delivered messages as `nonce_end - nonce_begin + 1` using plain `+` and `-` operators. When `nonce_begin` is 0 and `nonce_end` is `u64::MAX` (i.e., the full nonce space is delivered), the addition of `1` to `u64::MAX` overflows. In debug/test mode this panics; in release mode it wraps to 0, producing an incorrect reward. The codebase provides and consistently uses a `saturating_len()` method for the identical computation in `total_unrewarded_messages` and `DeliveredMessages::total_messages`, and has an explicit test (`total_unrewarded_messages_does_not_overflow`) for this exact scenario. The fix is to replace `nonce_end - nonce_begin + 1` with `(nonce_begin..=nonce_end).saturating_len()`.

This finding was found by Runtime Whitebox Fuzzer.

## Severity / Sensitivity

**low severity.** Public correctness report. The PoC demonstrates a correctness or robustness defect without a demonstrated direct confidentiality or privilege-escalation impact.

## Source Evidence

- [`bridges/primitives/messages/src/lib.rs:517-523`](https://github.com/paritytech/polkadot-sdk/blob/11eb7aa0d9f0aa1b5b42448705b9906c9fc8c7a9/bridges/primitives/messages/src/lib.rs#L517-L523) (source): The `calc_relayers_rewards` loop body uses `nonce_end - nonce_begin + 1` with plain arithmetic operators (line 521). When the clamped range is [0, u64::MAX], the `+ 1` causes an overflow panic. This is the exact line referenced in the failure output: 'panicked at bridges/primitives/messages/src/lib.rs:521:68: attempt to add with overflow'.
- [`bridges/primitives/messages/src/lib.rs:289-299`](https://github.com/paritytech/polkadot-sdk/blob/11eb7aa0d9f0aa1b5b42448705b9906c9fc8c7a9/bridges/primitives/messages/src/lib.rs#L289-L299) (source): `total_unrewarded_messages` computes the same inclusive-range length but uses `(front.messages.begin..=back.messages.end).saturating_len()`, demonstrating the codebase's intended pattern for overflow-safe range-length computation.
- [`bridges/primitives/messages/src/lib.rs:407-410`](https://github.com/paritytech/polkadot-sdk/blob/11eb7aa0d9f0aa1b5b42448705b9906c9fc8c7a9/bridges/primitives/messages/src/lib.rs#L407-L410) (source): `DeliveredMessages::total_messages` also uses `(self.begin..=self.end).saturating_len()` for the same computation, further confirming the established saturating-arithmetic pattern.
- [`bridges/primitives/messages/src/lib.rs:562-578`](https://github.com/paritytech/polkadot-sdk/blob/11eb7aa0d9f0aa1b5b42448705b9906c9fc8c7a9/bridges/primitives/messages/src/lib.rs#L562-L578) (existing-test): The test `total_unrewarded_messages_does_not_overflow` constructs UnrewardedRelayer entries with begin=0 and end=u64::MAX — the identical scenario that triggers the overflow in calc_relayers_rewards — and asserts the result is MessageNonce::MAX (saturated). This proves the codebase expects this range to be handled without panic.

## Self-contained PoC Unit Test

Place this complete file at `bridges/primitives/messages/src/generated_tests.rs` in a checkout of [`11eb7aa0d9f0aa1b5b42448705b9906c9fc8c7a9`](https://github.com/paritytech/polkadot-sdk/commit/11eb7aa0d9f0aa1b5b42448705b9906c9fc8c7a9) and ensure its existing parent test module includes the generated-test module.

```rust
use super::*;

use std::collections::BTreeMap;
use std::ops::RangeInclusive;

/// Compute expected per-relayer rewards using u128 intermediate arithmetic to avoid
/// overflow in `nonce_end - nonce_begin + 1`. Each relayer's reward equals the
/// saturating inclusive count of the intersection of their delivered range with
/// the received range.
fn reference_relayers_rewards<AccountId: Ord + Clone>(
	entries: &[UnrewardedRelayer<AccountId>],
	received_range: &RangeInclusive<MessageNonce>,
) -> BTreeMap<AccountId, MessageNonce> {
	let mut rewards: BTreeMap<AccountId, MessageNonce> = BTreeMap::new();
	for entry in entries {
		let nonce_begin = std::cmp::max(entry.messages.begin, *received_range.start());
		let nonce_end = std::cmp::min(entry.messages.end, *received_range.end());
		if nonce_end >= nonce_begin {
			let count = (nonce_end as u128 - nonce_begin as u128 + 1).min(u64::MAX as u128)
				as MessageNonce;
			let r = rewards.entry(entry.relayer.clone()).or_insert(0);
			*r = r.saturating_add(count);
		}
	}
	rewards
}

/// Verify that the canonical projection from `InboundLaneData` to
/// `UnrewardedRelayersState` is consistent, that `is_valid` accepts it, and that
/// any single-field perturbation is rejected.
fn assert_unrewarded_relayers_state_roundtrip<RelayerId>(lane: &InboundLaneData<RelayerId>) {
	let canonical = UnrewardedRelayersState::from(lane);
	assert!(canonical.is_valid(lane), "canonical state must validate");
	assert_eq!(
		canonical.unrewarded_relayer_entries,
		lane.relayers.len() as MessageNonce
	);
	assert_eq!(
		canonical.messages_in_oldest_entry,
		lane.relayers.front().map(|e| e.messages.total_messages()).unwrap_or(0)
	);
	assert_eq!(canonical.total_messages, lane.total_unrewarded_messages());
	assert_eq!(canonical.last_delivered_nonce, lane.last_delivered_nonce());

	for field in 0..4u8 {
		let mut p = canonical.clone();
		match field {
			0 =>
				p.unrewarded_relayer_entries = canonical.unrewarded_relayer_entries.wrapping_add(1),
			1 =>
				p.messages_in_oldest_entry = canonical.messages_in_oldest_entry.wrapping_add(1),
			2 => p.total_messages = canonical.total_messages.wrapping_add(1),
			_ => p.last_delivered_nonce = canonical.last_delivered_nonce.wrapping_add(1),
		}
		assert!(!p.is_valid(lane), "perturbed state must be rejected");
	}
}

/// Compute expected `(last_delivered_nonce, total_unrewarded_messages)` for an
/// `InboundLaneData` using an independent saturating oracle that never panics.
fn reference_inbound_lane_delivery<RelayerId>(
	lane: &InboundLaneData<RelayerId>,
) -> (MessageNonce, MessageNonce) {
	if lane.relayers.is_empty() {
		(lane.last_confirmed_nonce, 0)
	} else {
		let front = lane.relayers.front().unwrap();
		let back = lane.relayers.back().unwrap();
		let last_delivered = back.messages.end;
		let total = if back.messages.end < front.messages.begin {
			0
		} else {
			((back.messages.end as u128 - front.messages.begin as u128 + 1).min(u64::MAX as u128))
				as MessageNonce
		};
		(last_delivered, total)
	}
}

/// Verify that `maximal_incoming_message_size` equals `min((x / 3) * 2,
/// HARD_MESSAGE_SIZE_LIMIT)` computed with widened-integer reference arithmetic,
/// and never exceeds the hard cap.
fn assert_maximal_incoming_message_size_invariants(max_extrinsic_size: u32) {
	let result = maximal_incoming_message_size(max_extrinsic_size);
	let uncapped = ((max_extrinsic_size as u64) / 3 * 2) as u32;
	let reference = uncapped.min(HARD_MESSAGE_SIZE_LIMIT);
	assert_eq!(result, reference);
	assert!(result <= HARD_MESSAGE_SIZE_LIMIT);
	assert!(result <= uncapped);
}

/// Verify that `queued_messages` returns the correct inclusive range when
/// `latest_received_nonce` has a representable successor and does not exceed
/// `latest_generated_nonce`.
fn assert_outbound_queued_messages(lane: &OutboundLaneData) {
	if lane.latest_received_nonce < MessageNonce::MAX
		&& lane.latest_received_nonce <= lane.latest_generated_nonce
	{
		let range = lane.queued_messages();
		assert_eq!(*range.start(), lane.latest_received_nonce + 1);
		assert_eq!(*range.end(), lane.latest_generated_nonce);
	}
}

// RWF-RISKS: risk_eie3mfva, risk_nyvm4wzr
// RWF-INVARIANTS: inv_3zdpf4p3
#[ignore]
#[test]
fn test_risk_eie3mfva_full_width_intersection_overflow() {
	// A single relayer whose delivered range, after clamping to the received
	// range, spans the full u64 nonce space [0, u64::MAX]. The reward expression
	// `nonce_end - nonce_begin + 1` becomes `u64::MAX - 0 + 1`, which overflows.
	let entries = vec![UnrewardedRelayer {
		relayer: 1u64,
		messages: DeliveredMessages { begin: 0, end: MessageNonce::MAX },
	}];
	let received_range = 0..=MessageNonce::MAX;

	let expected = reference_relayers_rewards(&entries, &received_range);
	let actual = calc_relayers_rewards(entries.iter().cloned().collect(), &received_range);
	assert_eq!(actual, expected);
}

// RWF-RISKS: risk_cksnx6nn, risk_97fpht8k

// RWF-INVARIANTS: inv_nk4m9w78

// RWF-INVARIANTS: inv_95kmbsmd

// RWF-INVARIANTS: inv_ztw5ffrb

// RWF-INVARIANTS: inv_jdrzn699, inv_3a9hzqy6

// RWF-INVARIANTS: inv_2zuwz9pd
```

## Reproduction Command

Run from the root of the pinned `paritytech/polkadot-sdk` checkout:

```sh
cargo test -p bp-messages generated_tests::test_risk_eie3mfva_full_width_intersection_overflow -- --ignored --exact --nocapture
```

## Observed Result

```text
running 1 test
test generated_tests::test_risk_eie3mfva_full_width_intersection_overflow ... FAILED

failures:

failures:
    generated_tests::test_risk_eie3mfva_full_width_intersection_overflow

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 20 filtered out; finished in 0.00s
Finished `test` profile [unoptimized + debuginfo] target(s) in 0.64s
     Running unittests src/lib.rs (target/debug/deps/bp_messages-e062d0ee9907cbb0)

thread 'generated_tests::test_risk_eie3mfva_full_width_intersection_overflow' (2153380) panicked at bridges/primitives/messages/src/lib.rs:521:68:
attempt to add with overflow
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: test failed, to rerun pass `--lib`
```

## Expected Behavior

The operation must preserve the invariant described in the finding summary and source evidence. After the target defect is fixed, the focused ignored regression must pass without triggering the recorded assertion or state-integrity failure.

## Conclusion

Confirmed. The finding is well-supported by direct source code inspection. At bridges/primitives/messages/src/lib.rs:521, calc_relayers_rewards computes message count as `nonce_end - nonce_begin + 1` using plain arithmetic operators. When the clamped range is [0, u64::MAX], `u64::MAX - 0 + 1` overflows: it panics in debug mode (confirmed by the test output showing 'panicked at lib.rs:521:68: attempt to add with overflow') and silently wraps to 0 in release mode, producing an incorrect reward. The codebase already provides and uses `RangeInclusiveExt::saturating_len()` for the identical inclusive-range-length computation in two sibling functions: `InboundLaneData::total_unrewarded_messages` (lib.rs:295) and `DeliveredMessages::total_messages` (lib.rs:409). The existing test `total_unrewarded_messages_does_not_overflow` (lib.rs:562-578) constructs the exact [0, u64::MAX] scenario and asserts a saturated result of MessageNonce::MAX, proving the codebase expects this edge case to be handled without panic. The inconsistency in calc_relayers_rewards is a genuine defect. The function is called from production code (payment_adapter.rs:59) to compute relayer rewards during delivery confirmation. The generated test reproduced the panic, providing concrete validation evidence. No known related findings exist for deduplication.

## RWF Metadata

- Finding: `find_sy5vjapphg`
- Saved report: `freport_qx954chdqv`
- Run: `20260806-030112-5beaef`
- Target: `bp-messages`
- Source repository: `paritytech/polkadot-sdk`
- Source commit: [`11eb7aa0d9f0aa1b5b42448705b9906c9fc8c7a9`](https://github.com/paritytech/polkadot-sdk/commit/11eb7aa0d9f0aa1b5b42448705b9906c9fc8c7a9)
- Severity: `low`
- Category: `Integer overflow / unchecked arithmetic`

### PR DESCRIPTION
**Pull request by acatangiu**
`calc_relayers_rewards` counted messages as `end - begin + 1`, which overflows when the range covers the whole nonce space. Use the existing saturating_len() helper, and saturating_add for the running total.

Fixes https://github.com/paritytech/polkadot-sdk/issues/12907

### CODE REVIEW

File: bridges/primitives/messages/src/lib.rs [hunk 1]
```diff
@@ -577,6 +579,23 @@ mod tests {
...
+				relayer: 2,
+				messages: DeliveredMessages { begin: 2, end: MessageNonce::MAX },
+			},
+		]
+		.into_iter()
+		.collect();
+
+		let rewards = calc_relayers_rewards(relayers, &(0..=MessageNonce::MAX));
+		assert_eq!(rewards.get(&1), Some(&2));
+		assert_eq!(rewards.get(&2), Some(&(MessageNonce::MAX - 1)));
```
Comment (dr333ws) [inline bridges/primitives/messages/src/lib.rs]: These values don't overflow with the old code (`MAX - 2 + 1 = MAX - 1` fits in u64), so the test also passes without the fix. A full-range `0..=MAX` entry is needed to hit the `saturating_len` path, and a repeated relayer to hit the `saturating_add` path.

```suggestion
		let relayers = vec![
			UnrewardedRelayer {
				relayer: 1,
				messages: DeliveredMessages { begin: 0, end: MessageNonce::MAX },
			},
			UnrewardedRelayer {
				relayer: 1,
				messages: DeliveredMessages { begin: 0, end: MessageNonce::MAX },
			},
			UnrewardedRelayer { relayer: 2, messages: DeliveredMessages { begin: 0, end: 1 } },
		]
		.into_iter()
		.collect();

		let rewards = calc_relayers_rewards(relayers, &(0..=MessageNonce::MAX));
		assert_eq!(rewards.get(&1), Some(&MessageNonce::MAX));
		assert_eq!(rewards.get(&2), Some(&2));
```

**Review by dr333ws — APPROVED** [review]

**Review by bkchr — APPROVED** [review]
